File: data/code_pairs/fortran/1164.f90
SUBROUTINE sla_DC62S (V, A, B, R, AD, BD, RD)
    IMPLICIT NONE
    DOUBLE PRECISION V(6), A, B, R, AD, BD, RD
    DOUBLE PRECISION X, Y, Z, XD, YD, ZD, RXY2, RXY, R2, XYP

    X = V(1)
    Y = V(2)
    Z = V(3)
    XD = V(4)
    YD = V(5)
    ZD = V(6)

    RXY2 = X*X + Y*Y
    R2 = RXY2 + Z*Z

    IF (R2.EQ.0D0) THEN
        X = XD
        Y = YD
        Z = ZD
        RXY2 = X*X + Y*Y
        R2 = RXY2 + Z*Z
    END IF

    RXY = SQRT(RXY2)
    XYP = X*XD + Y*YD
    IF (RXY2.NE.0D0) THEN
        A = ATAN2(Y, X)
        B = ATAN2(Z, RXY)
        AD = (X*YD - Y*XD) / RXY2
        BD = (ZD*RXY2 - Z*XYP) / (R2*RXY)
    ELSE
        A = 0D0
        IF (Z.NE.0D0) THEN
            B = ATAN2(Z, RXY)
        ELSE
            B = 0D0
        END IF
        AD = 0D0
        BD = 0D0
    END IF

    R = SQRT(R2)
    IF (R.NE.0D0) THEN
        RD = (XYP + Z*ZD) / R
    ELSE
        RD = 0D0
    END IF
END SUBROUTINE sla_DC62S

File: data/code_pairs/cpp/1164.cpp
#include <cmath> // For sqrt(), atan2()

void sla_DC62S(const double V[6], double& A, double& B, double& R, double& AD, double& BD, double& RD) {
    double X, Y, Z, XD, YD, ZD, RXY2, RXY, R2, XYP;

    X = V[0];
    Y = V[1];
    Z = V[2];
    XD = V[3];
    YD = V[4];
    ZD = V[5];

    RXY2 = X * X + Y * Y;
    R2 = RXY2 + Z * Z;

    if (R2 == 0.0) {
        X = XD;
        Y = YD;
        Z = ZD;
        RXY2 = X * X + Y * Y;
        R2 = RXY2 + Z * Z;
    }

    RXY = sqrt(RXY2);
    XYP = X * XD + Y * YD;
    if (RXY2 != 0.0) {
        A = atan2(Y, X);
        B = atan2(Z, RXY);
        AD = (X * YD - Y * XD) / RXY2;
        BD = (ZD * RXY2 - Z * XYP) / (R2 * RXY);
    } else {
        A = 0.0;
        if (Z != 0.0) {
            B = atan2(Z, RXY);
        } else {
            B = 0.0;
        }
        AD = 0.0;
        BD = 0.0;
    }

    R = sqrt(R2);
    if (R != 0.0) {
        RD = (XYP + Z * ZD) / R;
    } else {
        RD = 0.0;
    }
}